### PR TITLE
smartcontract, examples: add API for Groth16 Verifier contract building and examples

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Three major improvements are included into this PR:
1. Add `zkpbinding` package that exports API to generate and interact with circuit-cpecific (and Groth-16 parameters-specific) Verifier Go smart contract. This API is assumed to be used by smart contract developers who wants to verify some ZK proofs on Neo.
2. Add an example of Verifier contract generation and proof generation/verification for a simple cubic circuit with the help of https://github.com/Consensys/gnark, `zkpbinding` package and `neotest`.
3. Port the https://github.com/neo-project/neo/issues/2647#issuecomment-1129849870 example from the C# to Go and invoke it with exactly the same proof and witnesses. This example is added to ensure that NeoGo BLS12-381 interoperability is compatible with the C# implementation. This example is not for the external users.

Close https://github.com/nspcc-dev/neo-go/issues/3002.

TODO:

- [x] Properly deal with scalar multipliers form.
- [x] Add clear and concise documentation.
- [x] Update interop deps (after PR finalization and all conversations resolving).

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
We have to think about moving Verifier contract auto-generation to a separate package so that contract devs are able to use it from the outside to generate their own verification contracts without massive copy-pasting. @roman-khimov, what do you think?

And with https://github.com/Consensys/gnark/pull/799, the generator-related code will be much simplier.

**Comment by AnnaShaleva** [comment]
@roman-khimov, could you, please, review the PR? We need to finalize the external smartcontract's generator API, let's fix the issues if there are any.

File: examples/zkp/cubic_circuit/main.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+import (
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/cs/r1cs"
+)
+
+// CubicCircuit defines a simple circuit x**3 + x + 5 == y
+// that checks that the prover knows the solution of the provided expression.
```
Comment (roman-khimov) [inline examples/zkp/cubic_circuit/main.go:24]: `for the`

File: examples/zkp/cubic_circuit/main.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		assignment = CubicCircuit{X: 3, Y: 35}
+	)
+
+	// Compile our circuit into a R1CS (a constraint system).
+	ccs, _ := frontend.Compile(ecc.BLS12_381.ScalarField(), r1cs.NewBuilder, &circuit)
+
+	// Once the circuit is compiled, you can run the three algorithms of a zk-SNARK back end:
+
+	// 1. One time setup (groth16 zkSNARK).
+	pk, vk, _ := groth16.Setup(ccs)
```
Comment (roman-khimov) [inline examples/zkp/cubic_circuit/main.go:66]: "Error handling omitted intentionally to simplify the example", worth mentioning somewhere.

File: pkg/smartcontract/zkpbinding/binding.go [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+	bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381"
+	"github.com/consensys/gnark-crypto/ecc/bn254/fr"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/backend/witness"
+	"github.com/nspcc-dev/neo-go/pkg/util/slice"
+)
+
+// Config represents a configuration for Verifier Go smart contract generator.
+type Config struct {
+	// VerifyingKey must be a Groth-16 BLS12-381 specific verifier key those
```
Comment (roman-khimov) [inline pkg/smartcontract/zkpbinding/binding.go:30]: `...key, parameters of which will be...

File: pkg/smartcontract/zkpbinding/binding.go [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+	// contract.
+	verifyCfg = `name: "Groth-16 Verifier contract"
+sourceurl: https://github.com/nspcc-dev/neo-go/
+supportedstandards: []`
+
+	// verifyGomod is a standard go.mod file containing module name, go version
+	// and dependency packages version needed for smart contract compilation.
+	verifyGomod = `module verify
+
+go 1.19
```
Comment (roman-khimov) [inline pkg/smartcontract/zkpbinding/binding.go:167]: Not yet.

File: pkg/smartcontract/zkpbinding/binding.go [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to generate go.mod file: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// byteSliceToStr is a codegen helper that converts byte slice to a go-like slice.
+func byteSliceToStr(s []byte) string {
```
Comment (roman-khimov) [inline pkg/smartcontract/zkpbinding/binding.go:313]: `%q` printf?

File: pkg/core/native/crypto.go [hunk 6]
```diff
@@ -231,6 +231,8 @@ func scalarFromBytes(bytes []byte, neg bool) (*fr.Element, error) {
 	}
 	// The input bytes are in the LE form, so we can't use fr.Element.SetBytesCanonical as far
 	// as it accepts BE.
+	// TODO: ensure that LE is really used in the C# node. More common way is to use BE form, and these
+	// lines of code are here only because that's the way how C# example works (https://github.com/neo-project/neo/issues/2647#issuecomment-1129849870).
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/native/crypto.go:235]: The check is preferable, not checked yet.

File: examples/zkp/xor_compat/go.mod [hunk 7]
```diff
@@ -0,0 +1,5 @@
+module verify
+
+go 1.19
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply examples/zkp/xor_compat/go.mod:3]: Fixed.

File: examples/zkp/xor_compat/verify.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
+// Package xor_compat contains an example of smart contract that uses BLS12-381
+// curves interoperability functionality to verify provided proof against provided
+// public input. This example is a full copy of C# smart contract presented in
+// https://github.com/neo-project/neo/issues/2647#issuecomment-1129849870 and is
+// aimed to check the compatibility of BLS12-381 interoperability functionality
+// between the NeoC# and NeoGo nodes. This example is circuit-specific and can
+// not be deployed into chain to verify other circuits.
+//
+// This example is constructed to proof that a&b=0 where a and b are private input.
+package xor_compat
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply examples/zkp/xor_compat/verify.go:16]: Fixed.

File: examples/zkp/xor_compat/verify.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	}
+)
+
+// VerifyProof verifies the given proof represented as three BLS12-381 points
+// against the public information represented as a list of serialized 32-bytes
+// field elements in the LE form. The verification process follows the GROTH-16
+// proving system and is taken from the
+// https://github.com/neo-project/neo/issues/2647#issuecomment-1002893109 without
+// changes. The verification process checks the following equality:
+// A*B = alpha*beta + sum(pub_input[i]*(beta*u_i(x)+alpha*v_i(x)+w_i(x))/gamma)*gamma + C*delta
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply examples/zkp/xor_compat/verify.go:46]: Fixed.

File: examples/zkp/xor_compat/verify.yml [hunk 10]
```diff
@@ -0,0 +1,3 @@
+name: "Groth16 proof compatibility example"
+sourceurl: https://github.com/nspcc-dev/neo-go/
+supportedstandards: []
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply examples/zkp/xor_compat/verify.yml:3]: Fixed.

File: examples/README.md [hunk 11]
```diff
@@ -20,19 +20,20 @@ You can use `my_wallet.json` to deploy example contracts.
...
+| [ZKP cubic circuit verification](zkp/cubic_circuit) | This example demonstrates how to create your own circuit and generate Groth-16 proof based on BLS12-381 elliptic curve points with the help of [consensys/gnark](https://pkg.go.dev/github.com/consensys/gnark). It also shows how to generate, deploy and invoke Verifier smart contract to verify proofs for the given circuit on the Neo chain with the help of [zkpbindings](https://pkg.go.dev/github.com/nspcc-dev/neo-go/pkg/smartcontract/zkpbinding) NeoGo package. The package also contains circuit tests implemented with [gnark/test](https://pkg.go.dev/github.com/consensys/gnark/test) to check the circuit validity and end-to-end proof generation/verification test implemented with [neotest](https://pkg.go.dev/github.com/nspcc-dev/neo-go/pkg/neotest) to demonstrate how to build, deploy and verify proofs via Verifier smart contract for the given circuit. |
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply examples/README.md:36]: Fixed.

File: examples/zkp/cubic_circuit/go.mod [hunk 12]
```diff
@@ -0,0 +1,66 @@
+module cubic_circuit
+
+go 1.19
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply examples/zkp/cubic_circuit/go.mod]: Fixed.

File: examples/zkp/cubic_circuit/main.go [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+// everything that developer needs to start using ZKP on the Neo platform with
+// Go SDK:
+//  1. The described cubic circuit implementation.
+//  2. The off-chain proof generation with the help of gnark-crypto library.
+//  3. The Go verification contract generation and deployment with the help of
+//     NeoGo library.
+//  4. The on-chain proof verification for various sets of input data (implemented
+//     as end-to-end test).
+//  5. A set of unit-tests aimed to check the circuit validity.
+package cubic_circuit
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply examples/zkp/cubic_circuit/main.go]: Fixed.

File: pkg/smartcontract/zkpbinding/binding.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/smartcontract/zkpbinding/binding.go]: Fixed.

File: pkg/smartcontract/zkpbinding/binding.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/smartcontract/zkpbinding/binding.go:30]: Fixed.

File: pkg/smartcontract/zkpbinding/binding.go (hunk 5 above)
Comment (AnnaShaleva) [reply pkg/smartcontract/zkpbinding/binding.go:313]: Fixed.

File: examples/zkp/cubic_circuit/main.go (hunk 2 above)
Comment (AnnaShaleva) [reply examples/zkp/cubic_circuit/main.go:66]: Fixed.

File: examples/zkp/cubic_circuit/main.go (hunk 1 above)
Comment (AnnaShaleva) [reply examples/zkp/cubic_circuit/main.go]: Fixed.

**Comment by AnnaShaleva** [comment]
Not ready for review yet, will test against the C# node and update.

**Comment by AnnaShaleva** [comment]
I've tested it on testnet (carefully with preliminary test `Verify` execution and result check). Here's the test:
```
func TestCubicCircuit_Compat(t *testing.T) {
	var (
		circuit    CubicCircuit
		assignment = CubicCircuit{X: 3, Y: 35}
	)

	// Compile our circuit into a R1CS (a constraint system).
	ccs, err := frontend.Compile(ecc.BLS12_381.ScalarField(), r1cs.NewBuilder, &circuit)
	require.NoError(t, err)

	// One time setup (groth16 zkSNARK).
	pk, vk, err := groth16.Setup(ccs)
	require.NoError(t, err)

	// Intermediate step: witness definition.
	witness, err := frontend.NewWitness(&assignment, ecc.BLS12_381.ScalarField())
	require.NoError(t, err)
	publicWitness, err := witness.Public()
	require.NoError(t, err)

	// Proof creation (groth16).
	proof, err := groth16.Prove(ccs, pk, witness)
	require.NoError(t, err)

	// Ensure that gnark can successfully verify the proof (just in case).
	err = groth16.Verify(proof, vk, publicWitness)
	require.NoError(t, err)

	// Now, when we're sure that the proof is valid, we can create and deploy verification
	// contract to the Neo testing chain.
	args, err := zkpbinding.GetVerifyProofArgs(proof, publicWitness)
	require.NoError(t, err)

	// Create contract file.
	tmpDir := t.TempDir()
	srcPath := filepath.Join(tmpDir, "verify.go")
	f, err := os.Create(srcPath)
	require.NoError(t, err)

	// Create contract configuration file.
	cfgPath := filepath.Join(tmpDir, "verify.yml")
	fCfg, err := os.Create(cfgPath)
	require.NoError(t, err)

	// Create contract go.mod and go.sum files.
	fMod, err := os.Create(filepath.Join(tmpDir, "go.mod"))
	require.NoError(t, err)
	fSum, err := os.Create(filepath.Join(tmpDir, "go.sum"))
	require.NoError(t, err)

	err = zkpbinding.GenerateVerifier(zkpbinding.Config{
		VerifyingKey: vk,
		Output:       f,
		CfgOutput:    fCfg,
		GomodOutput:  fMod,
		GosumOutput:  fSum,
	})
	require.NoError(t, err)

	require.NoError(t, f.Close())
	require.NoError(t, fCfg.Close())
	require.NoError(t, fMod.Close())
	require.NoError(t, fSum.Close())

	// rpcUrl := "https://rpc.t5.n3.nspcc.ru:20331"
	rpcUrl := "http://seed1t5.neo.org:20332"
	c, err := rpcclient.New(context.Background(), rpcUrl, rpcclient.Options{})
	require.NoError(t, err)
	require.NoError(t, c.Init())

	w, err := wallet.NewWalletFromFile("../../../n3testnet.json")
	require.NoError(t, err)
	acc := w.Accounts[0]
	require.NoError(t, acc.Decrypt("qwerty", w.Scrypt))

	s := transaction.Signer{
		Account: acc.ScriptHash(),
		Scopes:  transaction.CalledByEntry,
	}
	act, err := actor.New(c, []actor.SignerAccount{{
		Signer:  s,
		Account: acc,
	}})
	require.NoError(t, err)
	man := management.New(act)

	ctr := neotest.CompileFile(t, acc.ScriptHash(), srcPath, cfgPath)
	fmt.Println(ctr.Hash.StringLE())
	res, err := act.Wait(man.Deploy(ctr.NEF, ctr.Manifest, nil))
	require.NoError(t, err)
	require.Equal(t, vmstate.Halt, res.VMState, res.FaultException)

	res, err = act.Wait(act.SendCall(ctr.Hash, "verifyProof", args.A, args.B, args.C, args.PublicWitnesses))
	require.NoError(t, err)
	require.Equal(t, vmstate.Halt, res.VMState, res.FaultException)
	require.True(t, res.Stack[0].Value().(bool))
	fmt.Println(res.Container.StringLE())
}
```

Here's the contract: https://dora.coz.io/contract/neo3/testnet/c110099c025aadf10b5f4d6af32c65efd5268135
Here's the transaction (it pushes True on stack for both nodes): https://dora.coz.io/transaction/neo3/testnet/87a02195b611e4898275c2d21fb7d62ff60bef3a538bd8d54d47c7f20445cfec

So the `bls12381Mul` NeoGo implementation matches the NeoC# one.

File: pkg/smartcontract/zkpbinding/binding.go (hunk 5 above)
Comment (AnnaShaleva) [reply pkg/smartcontract/zkpbinding/binding.go:279]: Not fixed, we need it to be presented as a go byte slice, i.e. like this:
```
[]byte{176, 235, 189, 253, 2, 48, 68, 109, 189, 147, 187, 42, 110, 114, 152, 166, 152, 89, 224, 228, 23, 101, 228, 96, 51, 152, 165, 110, 244, 50, 156, 173, 189, 104, 135, 11, 8, 41, 156, 186, 120, 1, 70, 1, 232, 79, 147, 196}
```

File: examples/zkp/xor_compat/verify.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply examples/zkp/xor_compat/verify.go]: I mean leading spaces here, like 2 or 4 (how many does it need to become a monospace quote?)

File: pkg/smartcontract/zkpbinding/binding.go (hunk 5 above)
Comment (AnnaShaleva) [reply pkg/smartcontract/zkpbinding/binding.go:279]: What's the desired output here?

File: examples/zkp/xor_compat/verify.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply examples/zkp/xor_compat/verify.go]: image

Is it the desired behaviour?

File: pkg/core/native/crypto.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/native/crypto.go]: Checked, removed.

**Comment by AnnaShaleva** [comment]
The linter is broken with gnark, it's probably connected with some improper import, but I'm still searching how to fix this issue.

File: pkg/smartcontract/zkpbinding/binding.go [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+	}
+)
+
+// VerifyProof verifies the given proof represented as three serialized compressed
+// BLS12-381 points against the public information represented as a list of
+// serialized 32-bytes field elements in the LE form. Verification process
+// follows the Groth-16 proving system and is taken from the
+// https://github.com/neo-project/neo/issues/2647#issuecomment-1002893109 without
+// any changes. Verification process checks the following equality:
+// A*B = alpha*beta + sum(pub_input[i]*(beta*u_i(x)+alpha*v_i(x)+w_i(x))/gamma)*gamma + C*delta
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/smartcontract/zkpbinding/binding.go]: Or it's updated in subsequent commits, so not an issue.

File: examples/zkp/cubic_circuit/go.mod [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+	golang.org/x/sys v0.11.0 // indirect
+	golang.org/x/term v0.11.0 // indirect
+	golang.org/x/text v0.12.0 // indirect
+	golang.org/x/tools v0.11.1 // indirect
+	google.golang.org/protobuf v1.31.0 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
+	rsc.io/tmplfunc v0.0.3 // indirect
+)
+
+replace github.com/nspcc-dev/neo-go => ../../../
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply examples/zkp/cubic_circuit/go.mod]: Fixed.

**Comment by AnnaShaleva** [comment]
Linter is also fixed, the issue was caused by an update to gnark v0.9.0 in the core library. v0.9.0 requires go 1.19, and we still support go 1.18. Thus, https://github.com/nspcc-dev/neo-go/issues/3079 is left unclosed until we'll bump minimum required go version for neo-go.

**Comment by AnnaShaleva** [comment]
Dependencies are updated as far.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**